code/notificationbuilder/src/main/java/com/adobe/marketing/mobile/notificationbuilder/internal/builders/ProductRatingNotificationBuilder.kt

```diff
@@ -0,0 +1,226 @@
+        )
+
+        // check if confirm button needs to be shown
+        if (pushTemplate.ratingSelected >= 0) {
+            expandedLayout.setViewVisibility(R.id.rating_confirm, View.VISIBLE)
+            expandedLayout.setNotificationTitleTextColor(
+                pushTemplate.titleTextColor,
+                R.id.rating_confirm
+            )
+
+            // add pending intent for confirm click
+            // sticky is set to false as the notification will be dismissed after confirm click
+            val selectedRatingAction = pushTemplate.ratingActionList[pushTemplate.ratingSelected]
+            expandedLayout.setRemoteViewClickAction(
+                context,
+                trackerActivityClass,
+                R.id.rating_confirm,
+                selectedRatingAction.link,
+                pushTemplate.ratingSelected.toString(),
+                pushTemplate.tag,
+                false
+            )
+        } else {
+            // hide confirm if no rating is selected
+            expandedLayout.setViewVisibility(R.id.rating_confirm, View.INVISIBLE)
+        }
+
+        return notificationBuilder
+    }
+
+    /**
+     * Populates the rating icons in the notification.
+     *
+     * @param context the current [Context] of the application
+     * @param broadcastReceiverClass the [Class] of the broadcast receiver to set in the created pending intent
+     * @param expandedLayout the [RemoteViews] containing the expanded layout of the notification
+     * @param pushTemplate the [ProductRatingPushTemplate] object containing the product rating push template data
+     * @param packageName the `String` name of the application package used to locate the layout resources
+     * @param channelIdToUse the `String` containing the channel ID to use for the notification
+     */
+    private fun populateRatingIcons(
+        context: Context,
+        broadcastReceiverClass: Class<out BroadcastReceiver>?,
+        expandedLayout: RemoteViews,
+        pushTemplate: ProductRatingPushTemplate,
+        packageName: String?,
+        channelIdToUse: String
+    ) {
+        // set the rating icons in the notification based on the rating selected
+        for (i in 0 until pushTemplate.ratingActionList.size) {
+            val ratingIconLayout = RemoteViews(packageName, R.layout.push_template_product_rating_icon_layout)
+            val ratingIconImageView = R.id.rating_icon_image
+            if (i <= pushTemplate.ratingSelected) {
+                if (!ratingIconLayout.setRemoteViewImage(pushTemplate.ratingSelectedIcon, ratingIconImageView)) {
+                    throw NotificationConstructionFailedException("Image for selected rating icon is invalid.")
+                }
+            } else if (!ratingIconLayout.setRemoteViewImage(pushTemplate.ratingUnselectedIcon, ratingIconImageView)) {
+                throw NotificationConstructionFailedException("Image for unselected rating icon is invalid.")
+            }
```

Comment:
Good question. We should construct the notification only if both images are valid, will change the logic

Reply:
On second thought that might make the notification look weird if we use payload provided icon for unselected state and bundled icon for selected state. I think the best thing to do here is clearing the notification manually

Reply:
Yea, frankly I could not think of any other way to do this hence my question. Yea, given that this is a setup issue (icon being invalid) and because there is a way to remove the notification manually, I am inclined to leave it the way it is now.